dvc/remote/gdrive/__init__.py

```diff
@@ -0,0 +1,277 @@
+from __future__ import unicode_literals
+
+import os
+import posixpath
+import logging
+
+from funcy import cached_property
+from backoff import on_exception, expo
+
+from dvc.scheme import Schemes
+from dvc.path_info import CloudURLInfo
+from dvc.remote.base import RemoteBASE
+from dvc.config import Config
+from dvc.exceptions import DvcException
+from dvc.remote.gdrive.pydrive import (
+    RequestListFile,
+    RequestListFilePaginated,
+    RequestCreateFolder,
+    RequestUploadFile,
+    RequestDownloadFile,
+)
+from dvc.remote.gdrive.utils import FOLDER_MIME_TYPE
+
+logger = logging.getLogger(__name__)
+
+
+class GDriveURLInfo(CloudURLInfo):
+    @property
+    def netloc(self):
+        return self.parsed.netloc
+
+
+class RemoteGDrive(RemoteBASE):
+    scheme = Schemes.GDRIVE
+    path_cls = GDriveURLInfo
+    REGEX = r"^gdrive://.*$"
+    REQUIRES = {"pydrive": "pydrive"}
+    GDRIVE_USER_CREDENTIALS_DATA = "GDRIVE_USER_CREDENTIALS_DATA"
+    DEFAULT_USER_CREDENTIALS_FILE = ".dvc/tmp/gdrive-user-credentials.json"
+
+    def __init__(self, repo, config):
+        super(RemoteGDrive, self).__init__(repo, config)
+        self.no_traverse = False
+        self.path_info = self.path_cls(config[Config.SECTION_REMOTE_URL])
+        self.config = config
+        self.init_drive()
+
+    def init_drive(self):
+        self.gdrive_client_id = self.config.get(
+            Config.SECTION_GDRIVE_CLIENT_ID, None
+        )
+        self.gdrive_client_secret = self.config.get(
+            Config.SECTION_GDRIVE_CLIENT_SECRET, None
+        )
+        if not self.gdrive_client_id or not self.gdrive_client_secret:
+            raise DvcException(
+                "Please specify Google Drive's client id and "
+                "secret in DVC's config. Learn more at "
+                "https://man.dvc.org/remote/add."
+            )
+        self.gdrive_user_credentials_path = self.config.get(
+            Config.SECTION_GDRIVE_USER_CREDENTIALS_FILE,
+            self.DEFAULT_USER_CREDENTIALS_FILE,
+        )
+
+        self.root_id = self.get_path_id(self.path_info, create=True)
+        self.cached_dirs, self.cached_ids = self.cache_root_dirs()
+
+    @on_exception(expo, DvcException, max_tries=8)
+    def execute_request(self, request):
+        try:
+            result = request.execute()
+        except Exception as exception:
+            retry_codes = ["403", "500", "502", "503", "504"]
+            if any(code in str(exception) for code in retry_codes):
```

Comment:
I've discovered that catching all PyDrive's custom defined exceptions types doesn't help and some 403 "API limit usage exceeded" exceptions still are crashing execution. That's the reason to catch Exception. That seems the bug in PyDrive.

Reply:
@MaxRis is there a common class that connects them - something like HttpException at least? Could we analyze and see what types are being thrown?

Reply:
With all last changes the code looks as following (not yet published to GitHub and probably not all code review suggestion are applied yet):

```
    # 8 tries, start at 0.5s, multiply by golden ratio, cap at 10s
    @retry(8,
        errors=(GDriveRetriableError),
        timeout=lambda a: min(0.5 * 1.618 ** a, 10))
    def execute_request(self, request):
        from pydrive.files import ApiRequestError
        try:
            result = request.execute()
        except Exception as exception:
            retry_codes = ["403", "500", "502", "503", "504"]
            if any("HttpError {}".format(code) in str(exception) for code in retry_codes):
                raise GDriveRetriableError("Google API request failed")
            raise
        return result
```

It's turned out that PyDrive doesn't catch underlying `errors.HttpError` exceptions for `ListFile` API calls. This leads to 2 types of exceptions appearing: `googleapiclient.errors.HttpError` and `pydrive.files.ApiRequestError`. (Have created PR to PyDrive's upstream to fix this issue https://github.com/gsuitedevs/PyDrive/pull/180 )
`googleapiclient` is internal dependency of PyDrive and is not supposed to be used in DVC.
Only `googleapiclient.errors.HttpError` provides the way to retrieve http error code without exception's string parsing.

I believe till https://github.com/gsuitedevs/PyDrive/pull/180 is not get merged into upstream, catching top level `Exception` is only possible solution ( since it's not desired to reference `googleapiclient.errors.HttpError` from DVC ).

P.S. Also adjusted http error string lookup code to search for `HttpError ` text before error code itself.

Reply:
sounds reasonable, my only 2c in this is that since it's fragile a bit should we add some tests to detect the drift in case PyDrive changes the interface and this logic stops working?

Reply:
Catching `googleapiclient.errors.HttpError` doesn't sound that bad to me. Sure we ignore the abstraction, but we won't catch anything we don't want. So it's dirty, but safe.

Reply:
@Suor I guess that `googleapiclient` should be introduced as dependency as well?

Reply:
@MaxRis yes, with a comment, why do we need it.

Reply:
Done https://github.com/MaxRis/dvc/commit/783d8b610db9da2bdcee1817cd90366b435e81eb